Align Fortran arrays by calling posix_memalign

terminationshock opened the pull request:
Call `posix_memalign` to align arrays to the `n` byte boundary (usually used as 64 bytes on x86-64). Although the Intel Fortran compiler has a flag `-align array64byte`, which forces this alignment globally, nothing like this exists for the GNU compiler (and others). This implementation allows alignment independent of the compiler.

bhourahine commented on src/dftbp/common/memman.F90:
```diff
@@ -5,14 +5,64 @@
+    !> Number of array elements
+    integer :: size = 0
+
+    !> Chosen alignment in bytes (default: 64 byted)
```
```suggestion
    !> Chosen alignment in bytes (default: 64 bytes)
```

bhourahine commented on src/dftbp/common/memman.F90:
```diff
@@ -30,4 +80,59 @@ pure function incrmntOfArray(currentSize)
   end function incrmntOfArray
 
+
+  !> Allocate an array aligned to the n byte boundary (typically used for aligning to 64 byte)
```
```suggestion
  !> Allocate an array aligned to the n byte boundary (typically used for aligning to 64 bytes)
```

bhourahine commented on test/src/dftbp/unittests/common/memman.F90:
```diff
@@ -0,0 +1,73 @@
+!--------------------------------------------------------------------------------------------------!
+!  DFTB+: general package for performing fast atomistic simulations                                !
+!  Copyright (C) 2006 - 2022  DFTB+ developers group                                               !
```
```suggestion
!  Copyright (C) 2006 - 2023  DFTB+ developers group                                               !
```

aradi commented on src/dftbp/common/memman.F90:
```diff
@@ -30,4 +80,59 @@ pure function incrmntOfArray(currentSize)
+    @:ASSERT(size > 0)
+
+    if (present(alignment)) then
+      @:ASSERT(alignment >= dp)
```
You can not use `dp` as the number of bytes used to represent a given type, as it can have any compiler dependent positive value. (Many compilers choose 8 for doubles, but NAG uses 2 for example...). You could use `c_sizeof` instead.

aradi commented on src/dftbp/common/memman.F90:
```diff
@@ -5,14 +5,64 @@
+  type TAlignedArray
+
+    !> Actual array data
+    real(dp), pointer :: array(:)
```
```suggestion
    real(dp), pointer :: array(:) => null()
```

aradi commented on src/dftbp/common/memman.F90:
```diff
@@ -5,14 +5,64 @@
+    real(dp), pointer :: array(:)
+
+    !> Number of array elements
+    integer :: size = 0
```
I'd leave it away. User can query `size(instance%array)` to find out the size.

aradi commented on src/dftbp/common/memman.F90:
```diff
@@ -5,14 +5,64 @@
+    !> Chosen alignment in bytes (default: 64 bytes)
+    integer :: alignment = 64
+
+    !> Whether the array is allocated
+    logical :: isAllocated = .false.
```
Again, this is redundant IMO. Querying `associated(instance%array)` would do the same.

aradi commented on src/dftbp/common/memman.F90:
```diff
@@ -5,14 +5,64 @@
+    logical :: isAllocated = .false.
+
+    !> Pointer to the allocated memory
+    type(c_ptr) :: memoryPointer = c_null_ptr
```
Let's make it `private`, we should only use the Fortran pointer, unless if possible.

aradi commented on src/dftbp/common/memman.F90:
```diff
@@ -5,14 +5,64 @@
+
+  contains
+
+    procedure :: allocateAligned => TAlignedArray_allocateAligned
```
It is a matter of taste, and we can discuss, but I'd just call this `%allocate()`  instead of `%allocateAligned()`, as you can not allocate this array other than aligned.

aradi commented on src/dftbp/common/memman.F90:
```diff
@@ -30,4 +80,59 @@ pure function incrmntOfArray(currentSize)
+
+
+  !> Clean-up an aligned array by deallocating it
+  subroutine TAlignedArray_finalize(this)
```
Note: if someone creates an array of `type(TAlignedArray)` , the destructor will not be called when it goes out of scope! You should make it `elemental impure` to have it invoked for any ranks.

aradi reviewed (commented):
Thanks a lot! A few remarks in addition to the comments so far:
- We should guard the type again assignment, either by deallocating and reallocating on assignment (which would be my preference) or by just throwing an error.
-  We should consider to give only indirect access to the Fortran pointer (via a getter) to make sure, it can not be changed from outside (which would end up in a desaster).
- One could think to drop the storage of the C-ptr as one can regain it back using `c_loc()`, if needed (as for the finalization):

terminationshock commented:
@aradi I reworked the implementation. To make it even cleaner, I call `c_f_pointer` when `getArray` is invoked, so the caller always get a pointer, which can be safely nullified (see the test). The array remains allocated and will be freed when getting out of scope. I also added an assignment procedure that re-allocates and copies the data. Please feel free to criticize if you do not agree.

aradi commented on src/dftbp/common/memman.F90:
```diff
@@ -93,43 +90,76 @@ subroutine TAlignedArray_allocateAligned(this, size, alignment)
+
+
+  !> Fetch the pointer to the array data
+  subroutine TAlignedArray_getArray(this, array)
```
Yes, this is definitely the most robust solution. One stylistic remark (open for discussion): one could also implement this as function, so that on the caller side you had 
```
myArray => aligned%array()
```
instead of 
```
call aligned%getArray(myArray)
```.
I like the first one more from the optics, but it is more error prone, as if somebody writes `=` instead of `=>`, it will warrantedly crash at run-time. So, probably let's stay with the subroutine....

terminationshock replied:
I was also thinking about that, because we are using it in ELPA. However, I do not like it, because of the exact reason you have mentioned: If the user just writes a `=`, then everything gets messed up. The subroutine is safer for sure.

aradi commented on src/dftbp/common/memman.F90:
```diff
@@ -5,14 +5,61 @@
+    integer :: alignment = 64
+
+    !> Pointer to the allocated memory
+    type(c_ptr), private :: memoryPointer = c_null_ptr
```
I tend to add a `_` suffix to the private fields of a derived type.
```suggestion
    type(c_ptr), private :: memoryPointer_ = c_null_ptr
```

aradi reviewed (commented):
I like the solution a lot. I have only minor suggestions / remarks left.
